warning: output filename collision.

### Problem

Got lots of warnings like below
```
warning: output filename collision.
The lib target `cfg_if` in package `cfg-if v1.0.0` has the same output filename as the lib target `cfg_if` in package `cfg-if v1.0.0`.
Colliding filename is: target/debug/deps/libcfg_if-b07c5e93d2d15f7e.rlib
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
```

### Steps

no warnings:
- stable
- cargo 1.81.0-nightly (a515d4634 2024-07-02)

warnings:
- cargo 1.81.0-nightly (154fdac39 2024-07-07) start to pop up warnings.

### Possible Solution(s)

_No response_

### Notes

### Version

```text
nightly-x86_64-apple-darwin
```

The error can be reproduced with the follow steps in rust-lang/rust.

1. Check out https://github.com/rust-lang/rust/commit/53676730146e38e4697b6204c2ee61a9fd6b7e51
2. Set `~/.cargo/config.toml` with the follow configuration:
   ```toml
   [unstable]
   target-applies-to-host = true
   host-config = true
   ```
3. Run `./x.py` in the root of rust-lang/rust repo. The first invocation should pass.
4. Run again `./x.py`. This time it should fail with the output:
   ```
   Building bootstrap
   warning: output filename collision.
   The lib target `cc` in package `cc v1.0.97` has the same output filename as the lib target `cc` in package `cc v1.0.97`.
   Colliding filename is: /Users/user/dev/rust/build/bootstrap/debug/deps/libcc-0ac4eaa3c8400ccd.rlib
   The targets should have unique names.
   Consider changing their names to be unique or compiling them separately.
   This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
   warning: output filename collision.
   The lib target `cc` in package `cc v1.0.97` has the same output filename as the lib target `cc` in package `cc v1.0.97`.
   Colliding filename is: /Users/user/dev/rust/build/bootstrap/debug/deps/libcc-0ac4eaa3c8400ccd.rmeta
   The targets should have unique names.
   Consider changing their names to be unique or compiling them separately.
   This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
      Compiling cc v1.0.97
      Compiling lzma-sys v0.1.20
      Compiling xz2 v0.1.7
      Compiling bootstrap v0.0.0 (/Users/user/dev/rust/src/bootstrap)
   error[E0460]: found possibly newer version of crate `cc` which `cmake` depends on
      --> src/core/build_steps/llvm.rs:306:23
       |
   306 |         let mut cfg = cmake::Config::new(builder.src.join(root));
       |                       ^^^^^
       |
       = note: perhaps that crate needs to be recompiled?
       = note: the following crate versions were found:
               crate `cc`: /Users/user/dev/rust/build/bootstrap/debug/deps/libcc-0ac4eaa3c8400ccd.rmeta
               crate `cc`: /Users/user/dev/rust/build/bootstrap/debug/deps/libcc-0ac4eaa3c8400ccd.rlib
               crate `cmake`: /Users/user/dev/rust/build/bootstrap/debug/deps/libcmake-c88aa06d9a5ce765.rmeta
   
   error[E0463]: can't find crate for `cmake`
      --> src/core/build_steps/llvm.rs:594:15
       |
   594 |     cfg: &mut cmake::Config,
       |               ^^^^^ can't find crate
   
   error[E0463]: can't find crate for `cmake`
      --> src/core/build_steps/llvm.rs:817:77
       |
   817 | ...ection, cfg: &mut cmake::Config) {
       |                      ^^^^^ can't find crate
   
   error[E0463]: can't find crate for `cmake`
      --> src/core/build_steps/llvm.rs:896:23
       |
   896 | ...   let mut cfg = cmake::Config::new(builder.src.join("src/l...
       |                     ^^^^^ can't find crate
   
   error[E0463]: can't find crate for `cmake`
       --> src/core/build_steps/llvm.rs:1019:23
        |
   1019 |         let mut cfg = cmake::Config::new(&compiler_rt_dir);
        |                       ^^^^^ can't find crate
   
   Some errors have detailed explanations: E0460, E0463.
   For more information about an error, try `rustc --explain E0460`.
   error: could not compile `bootstrap` (lib) due to 5 previous errors
   failed to run: /Users/user/dev/rust/build/aarch64-apple-darwin/stage0/bin/cargo build --manifest-path /Users/user/dev/rust/src/bootstrap/Cargo.toml
   ```

If you run `./x.py` with stage0 cargo prior to this commit <https://github.com/rust-lang/cargo/commit/8457356c31a4cf7fd16d7ed03877645637cb0535>, it succeeds. So clearly there is something bad in 8457356c31a4cf7fd16d7ed03877645637cb0535.

Alright. Just narrowed bootstrap down to this real-world reproduction:

```toml
[package]
name = "cargo-14253"
version = "0.1.0"
edition = "2021"

[dependencies]
cc = "1.0.97"
lzma-sys = "0.1"

[profile.dev]
debug = 0
```

Run with

```
RUSTFLAGS="--cfg foo" cargo +beta -Zhost-config -Ztarget-applies-to-host build
```

and you'll get this compile error (it is not deterministic, sometimes succeeds):

```
warning: output filename collision.
The lib target `cc` in package `cc v1.1.7` has the same output filename as the lib target `cc` in package `cc v1.1.7`.
Colliding filename is: /home/user/projects/cargo-14253/target/debug/deps/libcc-abd516538b38f015.rlib
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
warning: output filename collision.
The lib target `cc` in package `cc v1.1.7` has the same output filename as the lib target `cc` in package `cc v1.1.7`.
Colliding filename is: /home/user/projects/cargo-14253/target/debug/deps/libcc-abd516538b38f015.rmeta
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
   Compiling libc v0.2.155
   Compiling cc v1.1.7
   Compiling pkg-config v0.3.30
   Compiling lzma-sys v0.1.20
error: failed to build archive at `/home/user/projects/cargo-14253/target/debug/deps/libcc-abd516538b38f015.rlib`: No such file or directory

error: could not compile `cc` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```

Note that `RUSTFLAGS="--cfg foo" and `[profile.dev] debug=0` are essential. They forced build (host) dependencies to share with normal dependencies.

---

For further investigation in Cargo here is a Cargo-flavored UI test reproduction:

<details><summary>click to expand</summary>
<p>

```rust
#[cargo_test]
fn host_config_shared_build_dep() {
    // rust-lang/cargo#14253
    Package::new("cc", "1.0.0").publish();
    Package::new("lzma-sys", "0.1.0")
        .file("src/lib.rs", "")
        .file("build.rs", "fn main() {}")
        .build_dep("cc", "1.0.0")
        .publish();
    let p = project()
        .file(
            "Cargo.toml",
            r#"
            [package]
            name = "bootstrap"
            edition = "2021"

            [dependencies]
            cc = "1.0.0"
            lzma-sys = "0.1.0"

            [profile.dev]
            debug = 0
            "#,
        )
        .file("src/lib.rs", "")
        .build();

    p.cargo("build")
        .masquerade_as_nightly_cargo(&["target-applies-to-host", "host-config"])
        .arg("-Zhost-config")
        .arg("-Ztarget-applies-to-host")
        .env("RUSTFLAGS", "--cfg foo")
        .with_status(101)
        // Sometimes it compiles. Not deterministic...
        .with_stderr_data(str![r#"
[UPDATING] `dummy-registry` index
[LOCKING] 3 packages to latest compatible versions
[DOWNLOADING] crates ...
[DOWNLOADED] lzma-sys v0.1.0 (registry `dummy-registry`)
[DOWNLOADED] cc v1.0.0 (registry `dummy-registry`)
[WARNING] output filename collision.
The lib target `cc` in package `cc v1.0.0` has the same output filename as the lib target `cc` in package `cc v1.0.0`.
Colliding filename is: [ROOT]/foo/target/debug/deps/libcc-[HASH].rlib
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
[WARNING] output filename collision.
The lib target `cc` in package `cc v1.0.0` has the same output filename as the lib target `cc` in package `cc v1.0.0`.
Colliding filename is: [ROOT]/foo/target/debug/deps/libcc-[HASH].rmeta
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
[COMPILING] cc v1.0.0
[ERROR] failed to build archive: No such file or directory

[ERROR] could not compile `cc` (lib) due to 1 previous error
"#])
        .run();
}
```

</p>
</details>

This is slightly tentative, but here's what I think is happening here so anyone can correct me if they know I'm wrong:

`-Ztarget-applies-to-host --config target-applies-to-host=false`\* make it possible to build two units (in this case cc) for the same output directory with different rustflags (one with artifact rustflags, one with none/host rust flags). Those flags aren't included in the hash which is used to disambiguate different versions of crates, resulting in a conflict. The actual error being produced\*\* appears to be the result of two invocations of `rustc` racing and clobbering eachother, but even when the error isn't produced we shouldn't expect the right behavior since two different artifacts have the same name and only one of them can sit there afterwards.

This wasn't caught because this is usually masked by the fact that the artifact (non-build-dep) version is passed either `-C debuginfo=2` in dev or `-C opt-level=3` in release builds. Both of which *are* included in the relevant hash.

I haven't invested time into verifying this yet, but I believe that this should only occur in cases where the behavior pre 8457356 was already wrong. Specifically as I currently understand it this only occurs when
 
 - no `--target` flag is passed. Otherwise host and target artifacts would be in different directories and we shouldn't hit this bug.
 - rustflags for non-host dependencies are passed, *or* with `-Zhost-config` rustflags for host-dependencies are passed (and if both are passed they are not identical). Otherwise the units would be deduplicated and we shouldn't hit this bug.
 - `-Ztarget-applies-to-host --config target-applies-to-host=false` (or equivalent) is passed, otherwise we just pass target rustflags to host deps and we wouldn't hit this bug.

And the combination of those bullet points mean that pre 8457356 we are passing the rustflags for host-deps to target artifacts (i.e. whatever is in the `host` table if `-Zhost-config` is set, or nothing if it isn't), which is very surprising (and wrong) behavior.

This obviously needs to be fixed, but if I'm correct that this should only occur in cases where we would otherwise pass the wrong rustflags I don't think reverting 8457356 is much of an improvement. Reverting would replace inconsistent wrong behavior which sometimes errors with consistent wrong behavior that never does.

I'm hopeful the fix is as simple as updating that hashing function and can be out quickly.

\* `-Zhost-config` is a substitute for `--config target-applies-to-host=false` because it changes the default for that config value to `false`.

\*\* Incidentally there's another variant of it, with `[ERROR] failed to build archive: No such file or directory` replaced with `[ERROR] failed to create encoded metadata from file: [NOT_FOUND]`.

<details>

<summary>
PS. Refined version of @weihanglo's test (`lzma-sys` crate as not necessary, `host-config` feature replace with `--config`)
</summary>

```rust
#[cargo_test]
fn host_config_shared_build_dep() {
    // rust-lang/cargo#14253
    Package::new("cc", "1.0.0").publish();
    let p = project()
        .file(
            "Cargo.toml",
            r#"
            [package]
            name = "bootstrap"
            edition = "2021"

            [dependencies]
            cc = "1.0.0"

            [build-dependencies]
            cc = "1.0.0"

            [profile.dev]
            debug = 0
            "#,
        )
        .file("src/lib.rs", "")
        .file("build.rs", "fn main() {}")
        .build();

    p.cargo("build")
        .masquerade_as_nightly_cargo(&["target-applies-to-host"])
        .arg("-Ztarget-applies-to-host")
        .arg("--config=target-applies-to-host=false")
        .env("RUSTFLAGS", "--cfg foo")
        .with_status(101)
        // Sometimes it compiles. Not deterministic...
        .with_stderr_data(str![[r#"
[UPDATING] `dummy-registry` index
[LOCKING] 2 packages to latest compatible versions
[DOWNLOADING] crates ...
[DOWNLOADED] cc v1.0.0 (registry `dummy-registry`)
[WARNING] output filename collision.
The lib target `cc` in package `cc v1.0.0` has the same output filename as the lib target `cc` in package `cc v1.0.0`.
Colliding filename is: [ROOT]/foo/target/debug/deps/libcc-[HASH].rlib
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
[WARNING] output filename collision.
The lib target `cc` in package `cc v1.0.0` has the same output filename as the lib target `cc` in package `cc v1.0.0`.
Colliding filename is: [ROOT]/foo/target/debug/deps/libcc-[HASH].rmeta
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
[COMPILING] cc v1.0.0
[ERROR] failed to build archive: No such file or directory

[ERROR] could not compile `cc` (lib) due to 1 previous error

"#]])
        .run();
}
```

</details>

--- a/src/cargo/core/compiler/build_runner/compilation_files.rs
+++ b/src/cargo/core/compiler/build_runner/compilation_files.rs
@@ -649,6 +649,26 @@ fn compute_metadata(
     // with user dependencies.
     unit.is_std.hash(&mut hasher);
 
+    // While we don't hash RUSTFLAGS because it may contain absolute paths that
+    // hurts reproducibility, we track whether a unit's RUSTFLAGS is from host
+    // config, so that we can generate a different metadata hash for runtime
+    // and compile-time units.
+    //
+    // HACK: This is a temporary hack for fixing rust-lang/cargo#14253
+    // Need to find a long-term solution to replace this fragile workaround.
+    // See https://github.com/rust-lang/cargo/pull/14432#discussion_r1725065350
+    if unit.kind.is_host() && !bcx.gctx.target_applies_to_host().unwrap_or_default() {
+        let host_info = bcx.target_data.info(CompileKind::Host);
+        let target_configs_are_different = unit.rustflags != host_info.rustflags
+            || unit.rustdocflags != host_info.rustdocflags
+            || bcx
+                .target_data
+                .target_config(CompileKind::Host)
+                .links_overrides
+                != unit.links_overrides;
+        target_configs_are_different.hash(&mut hasher);
+    }
+
     MetaInfo {
         meta_hash: Metadata(hasher.finish()),
         use_extra_filename: should_use_metadata(bcx, unit),

--- a/tests/testsuite/rustflags.rs
+++ b/tests/testsuite/rustflags.rs
@@ -1685,31 +1685,23 @@ fn host_config_shared_build_dep() {
         .masquerade_as_nightly_cargo(&["target-applies-to-host"])
         .arg("-Ztarget-applies-to-host")
         .arg("-Zhost-config")
-        // Sometimes it compiles. Not deterministic...
-        .with_stderr_data(str![[r#"
+        .with_stderr_data(
+            str![[r#"
 [UPDATING] `dummy-registry` index
 [LOCKING] 2 packages to latest compatible versions
 [DOWNLOADING] crates ...
 [DOWNLOADED] cc v1.0.0 (registry `dummy-registry`)
-[WARNING] output filename collision.
-The lib target `cc` in package `cc v1.0.0` has the same output filename as the lib target `cc` in package `cc v1.0.0`.
-Colliding filename is: [ROOT]/foo/target/debug/deps/libcc-[HASH].rlib
-The targets should have unique names.
-Consider changing their names to be unique or compiling them separately.
-This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
-[WARNING] output filename collision.
-The lib target `cc` in package `cc v1.0.0` has the same output filename as the lib target `cc` in package `cc v1.0.0`.
-Colliding filename is: [ROOT]/foo/target/debug/deps/libcc-[HASH].rmeta
-The targets should have unique names.
-Consider changing their names to be unique or compiling them separately.
-This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
 [COMPILING] cc v1.0.0
 [RUNNING] `rustc --crate-name cc [..]--cfg from_host[..]`
 [RUNNING] `rustc --crate-name cc [..]--cfg from_target[..]`
-[ERROR] failed to build archive: No such file or directory
+[COMPILING] bootstrap v0.0.0 ([ROOT]/foo)
+[RUNNING] `rustc --crate-name build_script_build [..]--cfg from_host[..]`
+[RUNNING] `[ROOT]/foo/target/debug/build/bootstrap-[HASH]/build-script-build`
+[RUNNING] `rustc --crate-name bootstrap[..]--cfg from_target[..]`
+[FINISHED] `dev` profile [unoptimized] target(s) in [ELAPSED]s
 
-[ERROR] could not compile `cc` (lib) due to 1 previous error
-
-"#]])
+"#]]
+            .unordered(),
+        )
         .run();
 }

--- a/tests/testsuite/rustflags.rs
+++ b/tests/testsuite/rustflags.rs
@@ -1642,3 +1642,74 @@ fn target_applies_to_host_rustdocflags_works() {
         )
         .run();
 }
+
+#[cargo_test]
+fn host_config_shared_build_dep() {
+    // rust-lang/cargo#14253
+    Package::new("cc", "1.0.0").publish();
+    let p = project()
+        .file(
+            "Cargo.toml",
+            r#"
+            [package]
+            name = "bootstrap"
+            edition = "2021"
+
+            [dependencies]
+            cc = "1.0.0"
+
+            [build-dependencies]
+            cc = "1.0.0"
+
+            [profile.dev]
+            debug = 0
+            "#,
+        )
+        .file("src/lib.rs", "")
+        .file("build.rs", "fn main() {}")
+        .file(
+            ".cargo/config.toml",
+            "
+            target-applies-to-host=false
+
+            [host]
+            rustflags = ['--cfg', 'from_host']
+
+            [build]
+            rustflags = ['--cfg', 'from_target']
+            ",
+        )
+        .build();
+
+    p.cargo("build -v")
+        .masquerade_as_nightly_cargo(&["target-applies-to-host"])
+        .arg("-Ztarget-applies-to-host")
+        .arg("-Zhost-config")
+        // Sometimes it compiles. Not deterministic...
+        .with_stderr_data(str![[r#"
+[UPDATING] `dummy-registry` index
+[LOCKING] 2 packages to latest compatible versions
+[DOWNLOADING] crates ...
+[DOWNLOADED] cc v1.0.0 (registry `dummy-registry`)
+[WARNING] output filename collision.
+The lib target `cc` in package `cc v1.0.0` has the same output filename as the lib target `cc` in package `cc v1.0.0`.
+Colliding filename is: [ROOT]/foo/target/debug/deps/libcc-[HASH].rlib
+The targets should have unique names.
+Consider changing their names to be unique or compiling them separately.
+This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
+[WARNING] output filename collision.
+The lib target `cc` in package `cc v1.0.0` has the same output filename as the lib target `cc` in package `cc v1.0.0`.
+Colliding filename is: [ROOT]/foo/target/debug/deps/libcc-[HASH].rmeta
+The targets should have unique names.
+Consider changing their names to be unique or compiling them separately.
+This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
+[COMPILING] cc v1.0.0
+[RUNNING] `rustc --crate-name cc [..]--cfg from_host[..]`
+[RUNNING] `rustc --crate-name cc [..]--cfg from_target[..]`
+[ERROR] failed to build archive: No such file or directory
+
+[ERROR] could not compile `cc` (lib) due to 1 previous error
+
+"#]])
+        .run();
+}

